### ISSUE TITLE
.skip().limit() performance is worse than .slice()

### ISSUE THREAD

**Issue by niieani** [rethinkdb/rethinkdb#6012]
To my understanding `.skip(10).limit(10)` is an equivalent of `.slice(10, 20)`. However, the performance of `slice` seems to be many times better. On a relatively small database I'm getting server time `2ms` for the query with slice() and `16ms` on the query with skip() + limit().

**Comment by danielmewes** [comment]
Thanks for reporting this @niieani .

Is this reproducible if you run the queries multiple times in a row (to make sure there's no caching artefacts)?

Are you on the latest RethinkDB version? (2.3.4)

**Comment by danielmewes** [comment]
This might be similar to https://github.com/rethinkdb/rethinkdb/issues/5155, though we did fix that particular issue.

**Comment by niieani** [comment]
Yes, it's reproducible. I get the same numbers every time. ~1-2ms for slice and 16ms for skip+limit.

Actually I just noticed that particular server is still on 2.3.1. 
I can try to reproduce in 2.3.4, but I don't have the time today to migrate.

**Comment by danielmewes** [comment]
Upgrading from 2.3.1 to 2.3.4 doesn't require migration.

However I don't think we changed anything in this logic, so it's likely that this is still an issue in 2.3.4.

**Comment by danielmewes** [comment]
@mlucy when you have a minute: any idea why `slice` might be different from `.skip.limit`?

**Comment by niieani** [comment]
@danielmewes I see. Well the difference in with my current database is hardly noticeable, although I did make the change in the code to use slice for now. Just reported cause it seemed odd to me :) Thanks for looking into it though!
Oh, by "migrate" I meant update the docker containers on production 🎱.

**Comment by mlucy** [comment]
@niieani --

* Do you see the same performance difference for arrays, or just for tables?
* How does the performance of `.limit(20).skip(10)` compare to `.skip(10).limit(10)` and `.slice(10, 20)`?

My best guess would be that we're calculating the batch size differently in those two situations, so one of them is reading more data than it needs to.

**Comment by niieani** [comment]
Actually, my initial report wasn't exact, I've done a bit more debugging and added the reversed order as suggested by @mlucy. Here are my findings:

slice(0, 20) - <1ms
skip(0).limit(20) - 1ms
limit(20).skip(0) - 1ms

slice(1000, 1020) - 1ms
skip(1000).limit(20) - 14ms
limit(20).skip(1000) - <1ms

slice(5000, 5020) - 1ms
skip(5000).limit(20) - 14ms
limit(20).skip(5000) - <1ms

Strangely, the reversed version, `limit().skip()` seems even faster than `slice` (repeatedly <1ms), although that might just be the error margin.

**Comment by niieani** [comment]
Forgot to answer the other question - I've only tested adding slice/skip-limit after `getAll(...)`.

**Comment by mlucy** [comment]
@niieani -- I think you want to write `.limit(5020).skip(5000)`, not `.limit(20).skip(5000)`.

If it's the same on an array as on a stream then it probably isn't the batching, though.

**Comment by niieani** [comment]
@mlucy I see.

In that case:

.limit(5020).skip(5000) - 1ms
.skip(5000).limit(20) - **69ms**
.slice(5000, 5020) - 1ms


I re-did my tests with the array, repeating it a number of times and I think that one time before was a bit unlucky. I think in the end array's work alright:

.coerceTo('array').limit(5020).skip(5000) - ~83-86ms
.coerceTo('array').skip(5000).limit(20) - ~83-85ms
.coerceTo('array').slice(5000, 5020) - ~82-86ms

**Comment by mlucy** [comment]
In that case it's almost certainly that the `skip` is ignoring the batchspec provided by the `limit`, so we're fetching more data than we need and throwing some of it away.  That shouldn't be super hard to fix.

Thanks for the help looking into this!

**Comment by analytik** [comment]
Running from Admin UI, a query like `r.db('foo').table('bar').orderBy('id', {index: 'id'}).slice(150000, 150100);` takes ~1.5s "roundtrip" time, and 30~40ms server time.

With `.skip(150000).limit(100)` it takes 2.3s roundtrip time, 2.1s server time.
